File: task3/src/main/java/com/kmnvxh222/task3/MainActivity.java
package com.kmnvxh222.task3;

import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.EditText;
import android.widget.TextView;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import java.util.ArrayList;
import java.util.List;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

public class MainActivity extends AppCompatActivity {

    private static List<Contacts> contacts = new ArrayList<>();
    private String TAG = "MainActivity";

    private RecyclerView recyclerView;
    private FloatingActionButton button;
    private ContactsRecyclerAdapter adapter;
    private TextView itemsNull;
    private EditText searchView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        recyclerView = findViewById(R.id.recyclerView);
        itemsNull = findViewById(R.id.itemsNull);
        searchView = findViewById(R.id.searchView);
        searchListener();

        adapter = new ContactsRecyclerAdapter(contacts);
        recyclerView.setAdapter(adapter);
        recyclerView.setLayoutManager(new LinearLayoutManager(this, RecyclerView.VERTICAL, false));

        adapter.setOnItemClickListener(new ContactsRecyclerAdapter.OnItemClickListener() {
            @Override
            public void onItemClick(View view, int position) {
                Intent intent = new Intent(MainActivity.this, EditContactActivity.class);
                intent.putExtra("EDIT_CONTACT", contacts.get(position));
                intent.putExtra("ID", contacts.get(position).getId());
                startActivityForResult(intent, 1000);
            }
        });

        button = findViewById(R.id.floatingActionButton);
        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(MainActivity.this, AddContactActivity.class);
                startActivityForResult(intent, 2000);
            }
        });
        itemsNull.setText(R.string.no_contacts);
    }

    //onActivityResult
    private void getData(@Nullable Intent data) {
        Contacts new_contact = (Contacts) data.getSerializableExtra("NEW_CONTACT");
        if(new_contact != null) {
            contacts.add(new_contact);
            adapter.notifyDataSetChanged();
        }
    }

    //onActivityResult
    private void editData(@Nullable Intent data) {
        Contacts editContact = (Contacts) data.getSerializableExtra("EDITED_CONTACT");
        String id = (String) data.getSerializableExtra("ID");

        if(editContact != null && id != null) {
            for(int i = 0; i < contacts.size(); i++) {
                if(contacts.get(i).getId().equals(id)) {
                    contacts.set(i, editContact);
                }
            }
            adapter.notifyDataSetChanged();
        }
    }

    private void searchListener() {
        searchView.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {
                search(s.toString());
            }
        });
    }

    void search(String text) {
        List<Contacts> temp = new ArrayList<>();
        for(Contacts c : contacts) {
            if(c.getName().contains(text)) {
                temp.add(c);
            }
        }
        adapter.updateList(temp);
    }

    //onActivityResult
    private void removeContact(@Nullable Intent data) {
        Integer position = (Integer) data.getSerializableExtra("REMOVE_CONTACT");
        if(position != null) {
            contacts.remove((int) position);
            adapter.notifyDataSetChanged();
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if(data != null) {
            if(requestCode == 1000) {
                removeContact(data);
                editData(data);
            }else if(requestCode == 2000) {
                getData(data);
                itemsNull.setText("");
            }
        }

    }
}

File: task5/src/main/java/com/kmnvxh222/task5/MainActivity.kt
package com.kmnvxh222.task5

import android.content.Intent
import android.database.sqlite.SQLiteDatabase
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.kmnvxh222.task5.ContactsRecyclerAdapter.OnItemClickListener
import com.kmnvxh222.task5.db.DBHelper
import kotlinx.android.synthetic.main.activity_main.floatingActionButton
import kotlinx.android.synthetic.main.activity_main.itemsNull
import kotlinx.android.synthetic.main.activity_main.recyclerView
import kotlinx.android.synthetic.main.activity_main.searchView

class MainActivity : AppCompatActivity() {

    private val contacts: MutableList<Contacts> = ArrayList()
    private lateinit var adapter: ContactsRecyclerAdapter
    private lateinit var db: SQLiteDatabase
    private lateinit var dbHelper: DBHelper

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        dataBaseInitialization()

        adapter = ContactsRecyclerAdapter(contacts)

        getAllContacts()

        recyclerView.adapter = adapter
        recyclerView.layoutManager = LinearLayoutManager(this, RecyclerView.VERTICAL, false)

        adapter.setOnItemClickListener(adapterClickListener)

        floatingActionButton.setOnClickListener { addContact() }

        searchListener()
    }

    private fun dataBaseInitialization() {
        dbHelper = DBHelper(this, null)
        db = dbHelper.readableDatabase
    }

    private val adapterClickListener = object : OnItemClickListener {
        override fun onItemClick(view: View, position: Int) {
            val intent = Intent(this@MainActivity, EditContactActivity::class.java)
            intent.putExtra("ID", contacts[position].id)
            startActivity(intent)
        }
    }

    private fun addContact() {
        val intent = Intent(this, AddContactActivity::class.java)
        startActivity(intent)
    }

    private fun getAllContacts() {
        contacts.clear()
        contacts.addAll(dbHelper.getAllContacts(db))
        if (contacts.isEmpty()) {
            itemsNull.setText(R.string.no_contacts)
        } else {
            itemsNull.text = ""
        }
        adapter.notifyDataSetChanged()
    }

    private fun searchListener() {
        searchView.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}

            override fun afterTextChanged(s: Editable) {
                search(s.toString())
            }
        })
    }

    fun search(text: String?) {
        val temp: MutableList<Contacts> = ArrayList()
        for (c in contacts) {
            if (c.name.contains(text.toString())) {
                temp.add(c)
            }
        }
        adapter.updateList(temp)
    }

    override fun onDestroy() {
        super.onDestroy()
        db.close()
    }

    override fun onResume() {
        super.onResume()
        getAllContacts()
        adapter.notifyDataSetChanged()
    }

}

